src/libsystemd/sd-device/device-monitor.c

```diff
@@ -627,8 +627,9 @@ _public_ int sd_device_monitor_filter_update(sd_device_monitor *m) {
         if (m->filter_uptodate)
                 return 0;
 
-        if (hashmap_isempty(m->subsystem_filter) &&
-            set_isempty(m->tag_filter)) {
+        if (m->snl.nl.nl_groups != MONITOR_GROUP_UDEV || (
```

Comment:
Yes, I figured that in the meantime, too. Sorry. I misunderstood MONITOR_GROUP_NONE.

Reply:
> Instead of refuse the filter here, may be it is better to refuse when filter is added in `sd_device_monitor_filter_add_match_xxx()`?

What benefit would that have? I considered it, but the way I did it,  I only need to insert a check in a single place. Moreover, `sd_device_monitor_filter_update()` is where the BPF logic is applied, so I thought it'd make sense to put it right there. People who read the function will immediately realize that the logic applies to "udev" monitors only.

Reply:
Why we need to accept filters for the unsupported group and silently ignore them? Setting filters for `MONITOR_GROUP_KERNEL` is definitely coding error. We should refuse and warn them. So, please add
```
assert_return(m->.snl.nl.nl_groups == MONITOR_GROUP_UDEV, -EINVAL);
```
to `sd_device_monitor_filter_add_match_xxx()`, and convert the condition you added to `assert()`.

Reply:
If you add a warning, you'll just confuse regular users of libudev who don't do anything out of the ordinary.

The application simply calls `udev_monitor_filter_add_match_subsystem()` and `udev_monitor_enable_receiving()`. The fact that this may cause a BPF filter to be created is an undocumented feature. libudev *does* support adding filters to kernel sockets, it just does the filtering in user space in that case.

If anything, the coding error is in libudev or in libsystemd itself. That's what my PR is trying to fix. You can of course move the check for `MONITOR_GROUP_KERNEL` into the callers, or never set `filter_uptodate = false` for kernel sockets in the first place. But the effect would be the same.
Or what am I missing?

Reply:
> libudev does support adding filters to kernel sockets, it just does the filtering in user space in that case.

libudev (or sd-device in libsystemd) should not support adding filters to kernel sockets, unfortunately currently they do.

> If anything, the coding error is in libudev or in libsystemd itself.

Adding filters to kernel socket is user's fault, and we should refuse it.

Reply:
> > libudev does support adding filters to kernel sockets, it just does the filtering in user space in that case.
> 
> libudev (or sd-device in libsystemd) should not support adding filters to kernel sockets, unfortunately currently they do.

Do you seriously plan to remove this functionality? It's a useful convenience functionality for users, even if for kernel sockets it's implemented less efficiently. It's likely to have quite a few users. The usage or non-usage of BPF is an implementation detail.

> > If anything, the coding error is in libudev or in libsystemd itself.
> 
> Adding filters to kernel socket is user's fault, and we should refuse it. "coding error" in my previous comment means there exist errors in _user's_ code.

I can't follow you. There is no documentation saying it's forbidden, and it works just fine (except for a minor performance overhead and the theoretical possiblity of false filtering, which my patch is supposed to fix). So why shouldn't applications use it this way? 

libudev has supported both BPF filtering and user-space filtering (`passes_filter()`)  since filtering functionality was first added in e14bdd8 in 2009. I don't know if it was ever intended to be used on kernel sockets, but as a matter of fact, libudev has silently accepted that for almost 12 years. Don't you think it's a bit late to declare this functionality deprecated now?

If you actually remove the functionality, apart from the confusion and frustration it's likely to cause among users of your library, the only effect will be that the callers will need to reimplement `passes_filter()` in their own code.

Anyway, I've provided this PR to fix what I believe is a bug in systemd. If you think otherwise, feel free to  close the PR and deal with this issue as you see fit.

Reply:
Ahhh, right now, I remembered that the filtering is done by two step BPF and `passes_filter()` (I forgot `passes_filter()`). Sorry, please forget my previous comments.

Yes, yes, you are right. And this looks good to me now.